streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -93,23 +136,118 @@ public void process(final K key, final V1 value) {
+                    //
+                    // the condition below allows us to process the late record without the need
+                    // to hold it in the temporary outer store
+                    if (!outerJoinWindowStore.isPresent() || timeTo < maxObservedStreamTime.get()) {
+                        context().forward(key, joiner.apply(key, value, null));
+                    } else {
+                        outerJoinWindowStore.ifPresent(store -> store.put(
+                            KeyAndJoinSide.make(isLeftJoin, key),
+                            LeftOrRightValue.make(isLeftJoin, value),
+                            inputRecordTimestamp));
+                    }
+                }
+            }
+        }
+
+        private void emitExpiredNonJoinedOuterRecords() {
+            outerJoinWindowStore.ifPresent(store ->
+                emitExpiredNonJoinedOuterRecords(store, recordWindowHasClosed));
+        }
+
+        private void emitExpiredNonJoinedOuterRecordsExcept(final K key, final long timestamp) {
+            outerJoinWindowStore.ifPresent(store -> {
+                final KeyAndJoinSide<K> keyAndJoinSide = KeyAndJoinSide.make(!isLeftJoin, key);
+
+                // Emit all expired records except the just found non-joined key. We need
+                // to emit all expired records before calling put(), otherwise the internal
+                // stream time will advance and may cause records out of the retention period to
+                // be deleted.
+                emitExpiredNonJoinedOuterRecords(store,
+                    recordWindowHasClosed
+                        .and(k -> !k.key().equals(keyAndJoinSide))
+                        .and(k -> k.window().start() != timestamp));
+
+                if (store.fetch(keyAndJoinSide, timestamp) != null) {
+                    // Delete the record. The previous emit call may not have removed this record
+                    // if the record window has not closed.
+                    store.put(keyAndJoinSide, null, timestamp);
+                }
+            });
+        }
+
+        @SuppressWarnings("unchecked")
+        private void emitExpiredNonJoinedOuterRecords(final WindowStore<KeyAndJoinSide<K>, LeftOrRightValue> store, final Predicate<Windowed<KeyAndJoinSide<K>>> emitCondition) {
+            try (final KeyValueIterator<Windowed<KeyAndJoinSide<K>>, LeftOrRightValue> it = store.all()) {
+                while (it.hasNext()) {
+                    final KeyValue<Windowed<KeyAndJoinSide<K>>, LeftOrRightValue> e = it.next();
+
+                    // Skip next records if the emit condition is false
+                    if (!emitCondition.test(e.key)) {
+                        break;
+                    }
+
+                    final K key = e.key.key().getKey();
+                    final To timestamp = To.all().withTimestamp(e.key.window().start());
+
+                    final R nullJoinedValue;
+                    if (isLeftJoin) {
```

Comment:
Do we need this (as we pass in a `ReversJoiner` in the "other" join processor anyway) ?

Reply:
Yes. The `ReverseJoiner` does not work well in the shared store scenario. I need cast the value and otherValue based on what join side we are. 

For instance, I would get an error if I remove the `(V1)` and `(V2)` casts. When using the casts, now I need to specify which value is the right one. In the left-side is `(V1) record.value.getLeftValue()`, in the right-side is `(V2) record.value.getLeftValue()`.
```
nullJoinedValue = joiner.apply(key,
                        (V1) record.value.getLeftValue(),
                        (V2) record.value.getRightValue());
```